pkg/storage/mvcc.go

```diff
@@ -4301,6 +4361,305 @@ func MVCCGarbageCollect(
 	return nil
 }
 
+// rangeKeyMergeTracker is tracking potential merges of range key fragments
+// when some of the versions are removed.
+type rangeKeyMergeTracker struct {
+	prevEndKey         roachpb.Key
+	previousTimestamps []hlc.Timestamp
+	ms                 *enginepb.MVCCStats
+}
+
+// update updates MVCCStats if previous range key fragments are adjacent to
+// current ones and remaining ones in current stack of fragments (indicated by
+// removed flag), match exactly with the previous stack (by their timestamps).
+func (t *rangeKeyMergeTracker) update(
+	startKey, endKey roachpb.Key, unsafeRangeKeys []MVCCRangeKeyValue,
+) {
+	if t.ms == nil {
+		return
+	}
+	if keyCount := len(unsafeRangeKeys); keyCount > 0 && keyCount == len(t.previousTimestamps) && t.prevEndKey.Equal(startKey) {
+		matching := true
+		for i, pts := range t.previousTimestamps {
+			if !pts.Equal(unsafeRangeKeys[i].RangeKey.Timestamp) {
```

Comment:
This needs to check that the values match too, since range keys with different values will not merge. For range tombstones, this can only happen when the `MVCCValueHeader` metadata differs.

Reply:
That's unfortunate as I'll have to clone value bytes into the "history" and then compare bytes there.
I think in this case value comparison should be done using raw bytes, not Value.EqualTagAndData since pebble doesn't interpret it.

Reply:
Yeah, we have to compare the raw byte value.